### DISCUSSION TITLE
EC11 Rotary Encoder on MCP23017 (pro micro) [different Pin Connection than example]

### DISCUSSION THREAD

**Discussion by nullscm** [tttapa/Control-Surface#961] [Q&A] [answered]
Hi :)

I know that you have an example for RotEncoders connected to a MCP but my pin configuration is a bit different so it's not working for me sadly.
The 5 Encoders are connected with 

```
ButtonEnc1 - 0
Enc1Turn - 1
Enc1Turn - 2 

ButtonEnc1 - 3
Enc1Turn - 4
Enc1Turn - 5 
...
```

already tried to configure them manually but it's not working:
```

#include <Control_Surface.h> // Include the Control Surface library
#include <Wire.h>
#include <AH/Hardware/ExtendedInputOutput/MCP23017.hpp>
#include <Arduino_Helpers.h>
using namespace ExtIO; // Use the extended input/output functions
using WireType = decltype(Wire); // The type of the I²C driver to use

USBDebugMIDI_Interface midi = 115200;

MCP23017<WireType> mcp {
  Wire, // The I²C driver to use
    0x21, // The I²C address offset (depends on the state of the address pins)
    1, // Optional: interrupt pin to detect input changes
};

CCRotaryEncoder enc {
  {mcp.pinA(1), mcp.pinA(2)}, // pins
    MCU::V_POT_1, // MIDI address (CC number + optional channel)
    1, // optional multiplier if the control isn't fast enough
};

void setup() {
  Wire.begin(); // Initialize the I²C interface
  Wire.setClock(800000); // Faster I²C clock
  mcp.begin(); // Initialize the MCP23017
  RelativeCCSender::setMode(relativeCCmode::MACKIE_CONTROL_RELATIVE);
  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); 
}
```

**Comment by tttapa** [comment]
Unfortunately, you cannot use the `ExtIO` elements like the `MCP23017` with encoders. This is deliberate: doing so would be too slow. I'd highly recommend connecting the encoders to the Arduino pins directly, preferably using interrupt-capable pins.

**Comment by nullscm** [comment]
Sadly that's not possible. But you already have an example for the MCP with 8 encoders, maybe you can hint me how to edit the class to use my pin configuration?

**Reply by tttapa** [reply, marked as answer]
It's probably easiest to use the `MCP23017` class, and then read the encoder pins manually using `digitalReadBuffered()`. Once you have the state of all encoder pins, you can pass it to [`RegistersEncoders::update()`](https://tttapa.github.io/Control-Surface-doc/Doxygen/d2/d7b/classAH_1_1RegisterEncoders.html). You can find some inspiration in https://tttapa.github.io/Control-Surface-doc/Doxygen/d6/dd2/Pin-Change-Interrupt-Encoders_8ino-example.html.

**Reply by nullscm** [reply]
alright thank you :)

**Comment by nullscm** [comment]
e.g. skipping pin 0,3,6,9,12,15 (the Button pins)
